### ISSUE TITLE
[Demonoid] & [TorrentDownload] can't be added if their sites can't be accessed - v0.14.243

### ISSUE THREAD

**Issue by ilike2burnthing** [Jackett/Jackett#7849]
Testing here done with v0.14.275, as well as v0.14.222 and v0.14.243 to confirm when change happened.

Tested with both current Docker container and clean Windows sandbox install.

I'm guessing this is as a result of https://github.com/Jackett/Jackett/commit/c78cd997ee1afa50c56f8ccc98c6eb68ec2f6d8b and https://github.com/Jackett/Jackett/commit/ef73523c8ee1ff87e0910c7abf5bb2ef94d683fd .

To reproduce:

1. remove Demonoid and/or TorrentDownload if you have them configured
2. disconnect/block from internet, alternatively use fake proxy details or remove logon details from real proxy
3. click Add Indexer
4. click `+` for Demonoid and/or TorrentDownload
5. get error and no new indexer in configured list

I tested all public indexers, the majority were added successfully, except for:
- Anidex
- Anime Tosho
- Corsaro.red
- Demonoid
- DivxTotal
- EpubLibre
- Horrible Subs
- Internet Archive
- MejorTorrent
- Newpct
- RARBG
- ShowRSS
- Solid Torrents
- TorrentDownload
- TorrentGalaxy.org
- Torrents.csv
- YTS

The rest of these fail in v0.14.222 as well, only Demonoid and TorrentDownload developed this behaviour with v0.14.243. That these are in the minority, I assume it's not the intended result. If there were a temporary issue with a site or for a user, they would be unable to add the indexer at all.

In some cases clicking the configuration wrench can help, e.g. choosing a proxy, adding a cookie, or entering a Captcha code, but those aren't options for most of the indexers listed.

I would assume this also affects some private and semi-private trackers, but I'm unable to test them.

**Comment by garfield69** [comment]
Yes. Any indexer that uses the cookie method will, if the config is added the first time,  need the cookie to be loaded in the config for the conficker to pass the test and be added to the active page.
The plus button should act like the wrench button for these indexes.
We may need to add logic behind the plus button if possible to detect if the indexer is using the cookie method and invoke the wrench anction instead of attempting to add the indexer to the active page.

**Comment by ilike2burnthing** [comment]
From what I can tell, the majority of indexers (public ones at least) aren't being tested before being added. Even with the internet disconnected, or blocked by fake proxy, I was able to add all but that list above.

**Comment by garfield69** [comment]
most public indexer do not have a login block in the yaml so there is no testing required and get them added to the active list.
but public indexer that have the login block (to support the cookie method session because the respective sites are using cloudflare or other challenges that prevent straight session access to the sites browser page) need to pass the test before they get added to the active list, hence the cause of the problem you have raised.

**Comment by ilike2burnthing** [comment]
Ah ok! Gotcha now.

So that explains Demonoid and TorrentDownload, they have the cookie method, and TGx, which has the Captcha challenge. However, the rest have no login requirement that I can see. They are all .cs files though. Is that why they have to pass the test first?

Feel free to change the title to something more fitting the plan to add logic behind the plus button for cookie/captcha detection 👍

**Comment by garfield69** [comment]
yes, the indexers written in c# use what ever custom method they choose to implement and the only reason the are written in c# in the first place is because there is something unique about the web site that the standard yaml code cannot support them.
mostly, c# are used for private sites (and so require a login), or any site that requires fetching results from an API rather than web scraping, or have multiple torrents per title, which the yaml cardigann processor cannot handle.
so c# all get tested before going on the active list.

**Comment by ngosang** [comment]
@ilike2burnthing I did some changes in the past. Could you test again?

**Comment by ilike2burnthing** [comment]
0.16.118.0 Windows Sandbox - simulating test failure/being unable to reach site, by using a fake proxy.

Same behaviour for all sites listed in OP - if test on clicking `Add` fails, they aren't added. All new public indexers since then can be added fine. Tested a handful of public indexers that were fine before, and they still get added without issue.

I didn't think that was going to change though, I thought it was remaining open for @garfield69 's comment,
```
We may need to add logic behind the plus button if possible to detect if the indexer is using the cookie method and invoke the wrench action instead of attempting to add the indexer to the active page.
```

If that's what you were asking about, then no, clicking `Add` for the above indexers closes the `Add indexer` and after a second or two an error displays. The configuration panel doesn't open.

Feel free to nab me for any other testing.

**Comment by ngosang** [comment]
@garfield69 I have been investigating and this problem can't be fixed in the UI, needs changes in the C# too. Meanwhile you can change public trackers with cookie to => semi-private (with 1 note in the yml or something).

**Comment by 101Dude** [comment]
Before I make a new issue. Is this issue what's causing TorrentGalaxy to fail here. I have FlareSolverr setup and it runs properly upon starting a search in my torrent client or using Jackett web UI. Then after a while I get a red check mark in Jackett UI for TorrentGalaxy. I can clear it by clicking on "Test" or going into "Configure/Wrench" and answering the captcha manually.

```xml
2023-11-29 08:08:14	Error	
Jackett.Common.IndexerException: Exception (torrentgalaxy): Relogin failed

[v0.21.1264.0] Jackett.Common.IndexerException: Exception (torrentgalaxy): Relogin failed
 ---> System.Exception: Parse error
 ---> System.Exception: Relogin failed
   at Jackett.Common.Indexers.CardigannIndexer.PerformQuery(TorznabQuery query) in ./Jackett.Common/Indexers/CardigannIndexer.cs:line 1719
   --- End of inner exception stack trace ---
   at Jackett.Common.Indexers.BaseWebIndexer.OnParseError(String results, Exception ex) in ./Jackett.Common/Indexers/BaseIndexer.cs:line 759
   at Jackett.Common.Indexers.CardigannIndexer.PerformQuery(TorznabQuery query) in ./Jackett.Common/Indexers/CardigannIndexer.cs:line 1878
   at Jackett.Common.Indexers.BaseIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in ./Jackett.Common/Indexers/BaseIndexer.cs:line 359
   --- End of inner exception stack trace ---
   at Jackett.Common.Indexers.BaseIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in ./Jackett.Common/Indexers/BaseIndexer.cs:line 359
   at Jackett.Common.Indexers.BaseWebIndexer.ResultsForQuery(TorznabQuery query, Boolean isMetaIndexer) in ./Jackett.Common/Indexers/BaseIndexer.cs:line 736
   at Jackett.Server.Controllers.ResultsController.Torznab(TorznabRequest request) in ./Jackett.Server/Controllers/ResultsController.cs:line 469
2023-11-29 08:08:14	Error	
Error on 20231129080814 for TorrentGalaxy.txtSystem.Exception: Relogin failed
   at Jackett.Common.Indexers.CardigannIndexer.PerformQuery(TorznabQuery query) in ./Jackett.Common/Indexers/CardigannIndexer.cs:line 1719




2023-11-29 08:08:14	Error	
CardigannIndexer (torrentgalaxy): Found captcha during automatic login, aborting
```

**Comment by garfield69** [comment]
>Is this issue what's causing TorrentGalaxy to fail here.

Insofar as TGx the website recognises that Jackett is not a browser and will occasionally require a captcha to be filled in.
The captcha answer is saved as a cookie which TGx will set a date expiry on, so this will expire eventually and require updating.
there is no point raising a separate issue for this, there is nothing Jackett can do about it.

**Comment by 101Dude** [comment]
@garfield69 I am no coder/programmer... excuse my ignorance. What is preventing Jackett from recognizing the cookie has expired or getting a new cookie on the next request after it's been blocked. FlareSolverr is working to get the initial cookie on startup when one, I am curious why Jackett is not asking FlareSolverr to request a new cookie?

**Comment by garfield69** [comment]
The website owns the cookie and creates whatever keys and values it needs for its own purposes. Same with cloudflare if the web site has enabled cloudflare security.
Some of the values are encrypted (session identifier, passwords, passports etc), and the life of a cookie is determined by the web site.
Neither flaresolverr or Jackett can manage the content of the cookies and or replace the values of the cookies, they both simply receive whatever the web site provides and saves them unaltered, and when Jackett makes a request to the web site to browser a page it provides the cookies that were saved.
IF the website for whatever reason decides the cookies are out of date or conflict with whatever criteria the web site decides then the web site will reject the session and ask for a login or security check afresh.

**Comment by 101Dude** [comment]
> IF the website for whatever reason decides the cookies are out of date or conflict with whatever criteria the web site decides then the web site will reject the session and ask for a login or security check afresh.

What I was driving at is that Jackett can "know" when the cookie has failed - it loses access to an indexer it previously had access too.  All I have to do is click on the "test" button and this error is cleared. It seems this could be automated with the right data saved and checked against:

date of last cookie acquired
date of last connection attempt ( did it pass or fail )
connection attempt failed = acquire new cookie ( do whatever clicking on "test" does)

It seems silly to have to go in and click on "test" everyday to request and receive a new cookie.

**Comment by ilike2burnthing** [comment]
If you check the logs when this happens, you'll see the error `CardigannIndexer (torrentgalaxy): Found captcha during automatic login, aborting`. This is important for private and semi-private indexers for which repeated failed login attempts (e.g. failing a captcha challenge) can get your account blocked or even banned. That won't be changing and has nothing to do with this issue.
